### ISSUE TITLE
PubSub: Subscriptions fail to work with PubSub emulator

### ISSUE THREAD

**Issue by obto** [GoogleCloudPlatform/google-cloud-node#2614]
Using google-cloud/pubsub: 0.14.1

I'm trying to get PubSub working for offline development by using the emulator and creating a dummy keyfile. This works perfectly for creating topics, subscriptions and publishing messages... however it fails whenever the subscription activates its connection pool. Looks like the connection pool attempts to fetch metadata on the subscription from Google's servers and topics don't...

Here is a script you can run for yourselves below to reproduce the problem:

```javascript
'use strict';

const PubSubBuilder = require('@google-cloud/pubsub');
const UTIL = require('util');
const PATH = require('path');

/*
dummy-gcloud-keyfile.json:
{
  "type": "service_account",
  "project_id": "dummy-account",
  "private_key_id": "<Dummy private key here>",
  "private_key": "<Dummy key here>",
  "client_email": "dummy-key@foobar.com",
  "client_id": "<Dummy client id>",
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://accounts.google.com/o/oauth2/token",
  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
  "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/dummy-key%40foobar.com"
}
*/

async function main() {
  // Listening to emulator on :8085
  const pubsub_client = PubSubBuilder({
    apiEndpoint: 'localhost:8085',
    projectId: 'foobar',
    keyFilename: PATH.resolve(__dirname, 'dummy-gcloud-keyfile.json'),
  });

  const topic = pubsub_client.topic('basic-test');
  const topic_exists = (await topic.exists())[0];
  if (!topic_exists) {
    console.log('Creating topic');
    await topic.create();
  }

  const subcription_name = 'basic-subscription';
  let subscription = topic.subscription(subcription_name);
  const subcription_exists = (await subscription.exists())[0];
  if (!subcription_exists) {
    console.log('Creating subscription');
    subscription =
        (await pubsub_client.createSubscription(topic, subcription_name))[0];
  }

  const publisher = topic.publisher();
  await publisher.publish(new Buffer('Test message'));
  console.log('Finished publishing messages');

  // NOTE: Everything works until here
  subscription.addListener('message', (message) => {
    console.log(message);
  });
}

main()
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });
```

**Comment by ctavan** [comment]
Related: #2610

**Comment by callmehiphop** [comment]
Sounds like maybe the emulator does not support the new streaming API.

@lukesneeringer @stephenplusplus do we know who is in charge of the PubSub emulator and/or if they support the new streaming interface?

**Comment by stephenplusplus** [comment]
Not sure myself. cc @jganetsk

**Comment by jganetsk** [comment]
Here's the official site for the emulator, with a link for reporting issues: https://cloud.google.com/pubsub/docs/emulator

**Comment by obto** [comment]
I am using the latest PubSub emulator available from the gcloud tool (2017.08.07)

Here's the output when running the script provided above:
```
node test-sample.js 
Creating topic
Creating subscription
Finished publishing messages
Auth error:Error: invalid_client
Auth error:Error: invalid_client
Auth error:Error: invalid_client
Auth error:Error: invalid_client
Auth error:Error: invalid_client
events.js:182
      throw er; // Unhandled 'error' event
      ^

Error: Getting metadata from plugin failed with error: invalid_client
    at ClientDuplexStream.<anonymous> (<path>/node_modules/@google-cloud/pubsub/src/connection-pool.js:218:21)
    at emitOne (events.js:115:13)
    at ClientDuplexStream.emit (events.js:210:7)
    at ClientDuplexStream._emitStatusIfDone (<path>/node_modules/grpc/src/node/src/client.js:260:10)
    at ClientDuplexStream._receiveStatus (<path>/node_modules/grpc/src/node/src/client.js:233:8)
    at <path>/node_modules/grpc/src/node/src/client.js:757:12
```

And the relevant emulator output:
```
[pubsub] INFO: Authentication interceptor: invalid or missing token
[pubsub] Sep 18, 2017 4:36:18 PM io.gapi.emulators.grpc.GrpcServer$3 operationComplete
[pubsub] INFO: Adding handler(s) to newly registered Channel.
[pubsub] Sep 18, 2017 4:36:18 PM io.gapi.emulators.netty.HttpVersionRoutingHandler channelRead
[pubsub] INFO: Detected HTTP/2 connection.
[pubsub] Sep 18, 2017 4:36:18 PM com.google.cloud.iam.testing.v1.shared.authorization.AuthInterceptor interceptCall
[pubsub] INFO: Authentication interceptor: Header value is null
[pubsub] Sep 18, 2017 4:36:18 PM com.google.cloud.iam.testing.v1.shared.authorization.AuthInterceptor interceptCall
```

**Comment by pongad** [comment]
Interesting. When running, I get a slightly different error:
```
Error: Resource not found (resource=basic-subscription).
```
The emulator is a little spammy, but I think the problem is with the NodeJS implementation, not the emulator. To test this, I created a Go program that subscribes and print messages:

```go
package main

import (
	"context"
	"log"

	"cloud.google.com/go/pubsub"
)

func main() {
	ctx := context.Background()
	c, err := pubsub.NewClient(ctx, "foobar") // Go autodetects the PUBSUB_EMULATOR_HOST variable
	if err != nil {
		log.Fatal(err)
	}

	sub := c.Subscription("basic-subscription")

	err = sub.Receive(ctx, func(cctx context.Context, msg *pubsub.Message) {
		log.Println(msg)
	})
	log.Fatal(err)
}
```
This program correctly prints out messages previously published by the node script. The `Receive` implementation uses streaming pull.

**Comment by callmehiphop** [comment]
Sorry about that, we ran into some issues with one of our core libraries and as a result had to authenticate the streamingPull calls separately.  It slipped my mind to add emulator support there.

Consequently I've opened a PR (#2617) that I believe addresses this. If you could test it to see if it resolves your issue, that would be a great help!

**Comment by obto** [comment]
LGTM. No further issues on my end :+1:

**Comment by callmehiphop** [comment]
I've released pubsub@0.14.2 which should resolve this issue. Thanks again for all the help!
